lib/mongodb.js

```diff
@@ -1881,14 +1881,25 @@ MongoDB.prototype.isObjectIDProperty = function(model, prop, value, options) {
   ) {
     return true;
   } else if ('string' === typeof value) {
+    // values doesn't look like an ObjectID
+    if (!value.match(/^[0-9a-fA-F]{24}$/)) return false;
```

Comment:
What happens when the property is configured with `mongodb: {dataType: 'ObjectID'}` byt the value doesn't look like an ObjectID? Will we store it as a string in MongoDB? IMO, such behavior is not correct, because the developer explicitly asked the values to be stored as ObjectIDs. Operations trying to set a non-ObjectID values to ObjectID properties should fail, or at least behave the same way as when setting non-ObjectID values to properties configured as `{type: ObjectID}`. Please add tests to verify.

Notice that when `prop.type === ObjectID` (lines 1880-1882 above), the value is not checked and the property is always considered as having ObjectID. IMO, we should use the same logic for `{type: ObjectID}`,

Reply:
On the second thought, I think the connector should always honor `mongodb: {dataType: 'ObjectID'}`, regardless of the actual value being set. I.e. the dataType check should be performed even for values that are not strings.

Reply:
> On the second thought, I think the connector should always honor `mongodb: {dataType: 'ObjectID'}`, regardless of the actual value being set. I.e. the dataType check should be performed even for values that are not strings.

If this means we _try_ to coerce them to ObjectID, then I am okay with it. If it's not an ObjectID like string, then coercion should throw an error, right?

Reply:
> If it's not an ObjectID like string, then coercion should throw an error, right?

Yes, that's what I would like to see. Here is my thinking:

- The application developer configured the model property to be stored as ObjectID in MongoDB.
- The client provided property value that's not a valid ObjectID.
- My opinion: the framework should reject the request as invalid.

IMO, when a developer configures as property as ObjectID, then the framework should ensure that only ObjectID value are ever stored in that property. That way the application developer can rely on the assumption that the string values will be always in ObjectID format and thus can keep their code simple (no need to check everywhere whether the string is really an ObjectID).